torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+                            cur_module)
+                        parent_qstate.mark_cur_op_complete(cur_module)
+
+                    elif hook_type is HookType.MODULE_IO_HOOKS:
+                        # TODO(future PR): add inputs io hook
+
+                        cur_qstate = cur_module._auto_quant_state
+                        cur_qstate.validate_is_at_first_idx()
+
+                        # original forward
+                        output = orig_module_call(self, *args, **kwargs)
+
+                        # after hooks
+                        assert isinstance(cur_qstate, AutoQuantizationState)
+                        output = cur_qstate.outputs_prepare_hook(
+                            output, first_call, qtensor_id)
+                        cur_qstate.validate_is_at_last_seen_idx()
+
+                    elif hook_type is HookType.ARG_DEQUANTS:
+                        output = orig_module_call(self, *args, **kwargs)
+                        # if this fp32 was inplace, make sure to set the output dtype
+                        # back to torch.float
+                        if hasattr(output, '_qtensor_info'):
+                            del output._qtensor_info
+
+                    else:
+                        output = orig_module_call(self, *args, **kwargs)
+
+                    if enable_logging:
+                        fqn = module_id_to_fqn.get(id(self), None)
+                        logger.debug(f"\nending fqn {fqn}")
+
+                    return output
+                finally:
+                    module_stack.pop()
+                    cur_module = old_module
+
+            torch.nn.Module.__call__ = record_module
+            torch.nn.Sequential.forward = _nn_sequential_patched_forward  # type: ignore[assignment]
+            nonlocal first_call
+            try:
+                # Create a list before iterating because we are adding new
+                # named modules inside the loop.
+                named_modules = list(self.named_modules())
+                for k, v in named_modules:
+
+                    # k is the global FQN, i.e. 'foo.bar.baz'
+                    # v is the module instance
+                    #
+                    # we need to associate the global FQN with SeenOp
+                    # for modules, this is the module FQN
+                    # for functions, this is the parent module FQN
+                    module_id_to_fqn[id(v)] = k
+
+                    has_qconfig = hasattr(v, 'qconfig') and v.qconfig is not None
+                    if has_qconfig and not is_leaf(v):
+                        if first_call:
+                            if v is self:
+                                # for the top level module only, specify input
+                                # and output dtypes
+                                v._auto_quant_state = AutoQuantizationState(
+                                    v.qconfig, input_dtypes, output_dtypes)
```

Comment:
Does this correspond to top level module of the full model? from the code it seems like we don't set the input_dtypes and output_dtypes anywhere (default to float). So wondering how we test other dtypes here.

Reply:
this is for handling any user defined dtypes for module boundaries.  Currently the only case of this is the default behavior that the top level model returns `torch.float` dtype for tensors.  In a future PR, we could allow the user to override this, similar to FX graph mode quantization.